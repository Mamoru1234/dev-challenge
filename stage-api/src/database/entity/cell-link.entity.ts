import {
  Column,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
  RelationId,
} from 'typeorm';
import { CellEntity } from './cell.entity';

@Entity('cell-link')
export class CellLinkEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column('uuid')
  @RelationId((it: CellLinkEntity) => it.fromCell)
  fromCellId: string;

  @ManyToOne(() => CellEntity, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  fromCell: CellEntity;

  @Column('uuid')
  @RelationId((it: CellLinkEntity) => it.toCell)
  toCellId: string;

  @ManyToOne(() => CellEntity, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  toCell: CellEntity;
}
