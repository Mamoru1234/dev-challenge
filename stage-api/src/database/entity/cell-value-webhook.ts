import {
  Column,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
  RelationId,
} from 'typeorm';
import { CellEntity } from './cell.entity';

@Entity('cell-value-webhook')
export class CellValueWebhookEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column('uuid')
  @RelationId((it: CellValueWebhookEntity) => it.cell)
  cellId: string;

  @ManyToOne(() => CellEntity, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  cell: CellEntity;

  @Column('text')
  url: string;
}
