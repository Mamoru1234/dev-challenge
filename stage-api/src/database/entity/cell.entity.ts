import { EquationNode } from 'equation-parser';
import {
  Column,
  Entity,
  Index,
  ManyToMany,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { ExternalValueEntity } from './external-value.entity';

@Entity('cell')
@Index(['cellId', 'sheetId'])
export class CellEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column('text')
  cellId: string;

  @Column('text')
  @Index()
  sheetId: string;

  @Column('text')
  value: string;

  @Column('text')
  result: string;

  @Column({
    type: 'jsonb',
    nullable: true,
  })
  equation?: EquationNode;

  @ManyToMany(() => ExternalValueEntity, (value) => value.cells)
  externals: ExternalValueEntity[];
}
