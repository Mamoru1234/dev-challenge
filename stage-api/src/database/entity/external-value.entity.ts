import {
  Column,
  Entity,
  Index,
  JoinTable,
  ManyToMany,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { CellEntity } from './cell.entity';

@Entity('external-value')
export class ExternalValueEntity {
  @PrimaryGeneratedColumn('increment')
  id: number;

  @Column('text')
  @Index({
    unique: true,
  })
  url: string;

  @Column('text')
  result: string;

  @ManyToMany(() => CellEntity, (value) => value.externals)
  @JoinTable()
  cells: CellEntity[];
}
