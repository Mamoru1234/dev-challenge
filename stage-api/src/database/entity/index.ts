import { CellLinkEntity } from './cell-link.entity';
import { CellValueWebhookEntity } from './cell-value-webhook';
import { CellEntity } from './cell.entity';

export const ENTITIES = [CellEntity, CellLinkEntity, CellValueWebhookEntity];
