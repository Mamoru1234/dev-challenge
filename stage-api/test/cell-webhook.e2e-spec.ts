import { HttpStatus, INestApplication } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { AppModule } from '../src/app.module';
import { enableValidation } from '../src/app.utils';
import { getDataSourceToken } from '@nestjs/typeorm';
import { DataSource, Repository } from 'typeorm';
import * as request from 'supertest';
import { CellValueWebhookEntity } from '../src/database/entity/cell-value-webhook';
import { pick } from 'lodash';

describe('Cell webhooks', () => {
  let app: INestApplication;
  let cellValueWebhookRepository: Repository<CellValueWebhookEntity>;

  const client = {
    subscribe: (sheetId: string, cellId: string, body: any) =>
      request(app.getHttpServer())
        .post(`/${sheetId}/${cellId}/subscribe`)
        .send(body),
    createCell: (sheetId: string, cellId: string, value: string) =>
      request(app.getHttpServer())
        .post(`/${sheetId}/${cellId}`)
        .send({ value }),
  };

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    enableValidation(app);
    await app.init();
    const dataSource: DataSource = app.get(getDataSourceToken('default'));
    cellValueWebhookRepository = dataSource.getRepository(
      CellValueWebhookEntity,
    );
    await dataSource.query('TRUNCATE TABLE cell CASCADE;');
  });

  afterEach(async () => {
    app?.close();
  });

  describe('POST /:sheet_id/:cell_id/subscribe', () => {
    describe('invalid request', () => {
      it('no webhook url', async () => {
        const { body } = await client
          .subscribe('test', 'test', {})
          .expect(HttpStatus.BAD_REQUEST);
        expect(body).toMatchSnapshot();
      });
      it('webhook is invalid url', async () => {
        const { body } = await client
          .subscribe('test', 'test', {
            webhook_url: 'test',
          })
          .expect(HttpStatus.BAD_REQUEST);
        expect(body).toMatchSnapshot();
      });
      it('url without protocol', async () => {
        const { body } = await client
          .subscribe('test', 'test', {
            webhook_url: '//www.google.com',
          })
          .expect(HttpStatus.BAD_REQUEST);
        expect(body).toMatchSnapshot();
      });
      it('url without https', async () => {
        const { body } = await client
          .subscribe('test', 'test', {
            webhook_url: 'http://www.google.com',
          })
          .expect(HttpStatus.BAD_REQUEST);
        expect(body).toMatchSnapshot();
      });
      it('no cell', async () => {
        await client
          .subscribe('test', 'test', {
            webhook_url: 'https://www.google.com',
          })
          .expect(HttpStatus.NOT_FOUND);
      });
    });

    it.only('should create subscription', async () => {
      await client
        .createCell('test', 'test', 'aTestValue')
        .expect(HttpStatus.CREATED);
      await client
        .subscribe('test', 'test', {
          webhook_url: 'https://www.google.com',
        })
        .expect(HttpStatus.CREATED);
      const webhooks = await cellValueWebhookRepository.find();
      expect(webhooks.map((it) => pick(it, 'url'))).toMatchSnapshot();
    });
  });
});
