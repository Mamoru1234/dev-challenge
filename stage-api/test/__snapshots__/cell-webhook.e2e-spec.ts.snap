// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell webhooks POST /:sheet_id/:cell_id/subscribe invalid request no webhook url 1`] = `
{
  "error": "Bad Request",
  "message": [
    "webhookUrl must be shorter than or equal to 1024 characters",
    "webhookUrl must be a URL address",
    "webhookUrl must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`Cell webhooks POST /:sheet_id/:cell_id/subscribe invalid request url without https 1`] = `
{
  "error": "Bad Request",
  "message": [
    "webhookUrl must be a URL address",
  ],
  "statusCode": 400,
}
`;

exports[`Cell webhooks POST /:sheet_id/:cell_id/subscribe invalid request url without protocol 1`] = `
{
  "error": "Bad Request",
  "message": [
    "webhookUrl must be a URL address",
  ],
  "statusCode": 400,
}
`;

exports[`Cell webhooks POST /:sheet_id/:cell_id/subscribe invalid request webhook is invalid url 1`] = `
{
  "error": "Bad Request",
  "message": [
    "webhookUrl must be a URL address",
  ],
  "statusCode": 400,
}
`;

exports[`Cell webhooks POST /:sheet_id/:cell_id/subscribe should create subscription 1`] = `
[
  {
    "url": "https://www.google.com",
  },
]
`;
